src/main/java/org/openrewrite/java/spring/boot3/MigrateWebMvcTagsToObservationConvention.java

```diff
@@ -0,0 +1,291 @@
+                                //noinspection DataFlowIssue
+                                return md.withBody(md.getBody().withStatements(ListUtils.insertAll(md.getBody().getStatements(), md.getBody().getStatements().size() - 1, getTagsBodyStatements)));
+                            }
+                            return stmt;
+                        })));
+                maybeAddImport(DEFAULTSERVERREQUESTOBSERVATIONCONVENTION_FQ);
+                maybeAddImport(KEYVALUE_FQ);
+                maybeAddImport(KEYVALUES_FQ);
+                maybeAddImport(SERVERREQUESTOBSERVATIONCONVENTION_FQ);
+                maybeRemoveImport(HTTPSERVLETREQUEST_FQ);
+                maybeRemoveImport(HTTPSERVLETRESPONSE_FQ);
+                maybeRemoveImport(TAG_FQ);
+                maybeRemoveImport(TAGS_FQ);
+                maybeRemoveImport(WEBMVCTAGS_FQ);
+                maybeRemoveImport(WEBMVCTAGSPROVIDER_FQ);
+                updateCursor(newClassDeclaration);
+                return (J.ClassDeclaration) super.visitClassDeclaration(newClassDeclaration, ctx);
+            }
+
+            @Override
+            public J.MethodDeclaration visitMethodDeclaration(J.MethodDeclaration method, ExecutionContext ctx) {
+                J.MethodDeclaration m = (J.MethodDeclaration) super.visitMethodDeclaration(method, ctx);
+                J.VariableDeclarations methodParam = (J.VariableDeclarations) m.getParameters().get(0);
+                J.Identifier methodParamIdentifier = methodParam.getVariables().get(0).getName();
+                Boolean addHttpServletResponse = getCursor().pollMessage("addHttpServletResponse");
+                Boolean addHttpServletRequest = getCursor().pollMessage("addHttpServletRequest");
+                if (Boolean.TRUE.equals(addHttpServletResponse) && m.getBody() != null) {
+                    m = JavaTemplate.builder("HttpServletResponse response = #{any()}.getResponse();")
+                            .imports(HTTPSERVLETRESPONSE_FQ, SERVERREQUESTOBSERVATIONCONVENTION_FQ)
+                            .javaParser(JavaParser.fromJavaVersion().classpathFromResources(ctx, "jakarta.servlet-api", "spring-web-6.+", "micrometer-commons", "micrometer-observation"))
+                            .build()
+                            .apply(updateCursor(m), m.getBody().getCoordinates().firstStatement(), methodParamIdentifier);
+                }
+                if (Boolean.TRUE.equals(addHttpServletRequest) && m.getBody() != null) {
+                    m = JavaTemplate.builder("HttpServletRequest request = #{any()}.getCarrier();")
+                            .imports(HTTPSERVLETREQUEST_FQ, SERVERREQUESTOBSERVATIONCONVENTION_FQ, "io.micrometer.observation.transport.ReceiverContext")
+                            .javaParser(JavaParser.fromJavaVersion().classpathFromResources(ctx, "jakarta.servlet-api", "spring-web-6.+", "micrometer-commons", "micrometer-observation"))
```

Comment:
Here you reference `micrometer-observation` but I don't see any matching jar in `/src/main/resources/META-INF/rewrite/classpath` or `build.gradle.kts` > `parserClasspath("...")`. Could you explain that part to me?

Reply:
Most likely it's transitively provided by spring-web, I've looked at these two templates again and was able to reduce the imports and required classpath, but micrometer-observations remains, would you like me to explicitly add it?